dali/kernels/common/scatter_gather.h

```diff
@@ -116,16 +103,95 @@ class DLL_PUBLIC ScatterGatherGPU {
    */
   void MakeBlocks();
 
+  size_t max_size_per_block_ = kDefaultBlockSize;
+  std::vector<CopyRange> blocks_;
+  size_t block_capacity_ = 0;
+  size_t size_per_block_ = 0;
+};
+
+
+/**
+ * Implements a device-to-device batch copy of multiple sources to multiple destinations
+ */
+class DLL_PUBLIC ScatterGatherGPU : public ScatterGatherBase {
+ public:
+  static constexpr size_t kDefaultBlockSize = 64<<10;
+
+  ScatterGatherGPU() = default;
+
+  ScatterGatherGPU(size_t max_size_per_block, size_t estimated_num_blocks)
+      : ScatterGatherBase(max_size_per_block, estimated_num_blocks) {
+    ReserveGPUBlocks();
+  }
+
+  explicit ScatterGatherGPU(size_t max_size_per_block) : ScatterGatherBase(max_size_per_block) {
+    ReserveGPUBlocks();
+  }
+
+  ScatterGatherGPU(size_t max_size_per_block, size_t total_size, size_t num_ranges)
+      : ScatterGatherBase(max_size_per_block, total_size, num_ranges) {
+    ReserveGPUBlocks();
+  }
+
+  /**
+   * @brief Executes the copies
+   * @param stream     - the cudaStream on which the copies are scheduled
+   * @param reset      - if true, calls Reset after processing is over
+   * @param method     - see ScatterGatherGPU::CopyMethod
+   * @param memcpyKind - determines the cudaMemcpyKind when using cudaMemcpy
+   */
+  DLL_PUBLIC void
+  Run(cudaStream_t stream, bool reset = true, Method method = Method::Default,
+      cudaMemcpyKind memcpyKind = cudaMemcpyDefault);
+
+  using CopyRange = detail::CopyRange;
+
+ private:
   /**
    * @brief Reserves GPU memory for the description of the blocks.
    */
   void ReserveGPUBlocks();
 
-  size_t max_size_per_block_ = kDefaultBlockSize;
-  std::vector<CopyRange> blocks_;
   kernels::memory::KernelUniquePtr<CopyRange> blocks_dev_;
```

Comment:
When you're at it - perhaps use `DeviceBuffer<CopyRange>`? Then you could allocate and copy it with simple `blocks_dev_.from_host(blocks_);` - it would handle buffer growth and all other stuff.

Reply:
done